components/core/src/crypto/hash.rs

```diff
@@ -27,34 +29,57 @@ const HASH_DIGEST_SIZE: usize = libsodium_sys::crypto_generichash_BYTES as usize
 /// or parse it from a hex string.
 #[derive(Clone, Debug)]
 pub struct Blake2bHash {
-    digest:     [u8; HASH_DIGEST_SIZE],
-    /// Temporary field to support Deref<str> for backwards
-    /// compatibility with Builder until it can use the new types.
-    hex_string: String,
+    digest: [u8; HASH_DIGEST_SIZE],
 }
 
 impl Blake2bHash {
-    /// Temporary constructor while we store the hex encoding in the
-    /// type directly.
-    fn new(digest: [u8; HASH_DIGEST_SIZE]) -> Self {
-        let hex_string = hex::encode(&digest).to_lowercase();
-        Blake2bHash { digest, hex_string }
+    /// Calculate the BLAKE2b hash of the contents of a file.
+    pub fn from_file<P>(filename: P) -> Result<Self>
+        where P: AsRef<Path>
+    {
+        let file = File::open(filename.as_ref())?;
+        let mut reader = BufReader::new(file);
+        Self::from_reader(&mut reader)
+    }
+
+    /// Calculate the BLAKE2b hash of a sequence of bytes.
+    pub fn from_bytes<B>(data: B) -> Self
+        where B: AsRef<[u8]>
+    {
+        let mut state = hash_state();
+        state.update(data.as_ref());
+        state.finalize().into()
+    }
+
+    /// Calculate the BLAKE2b hash of a Read implentation.
+    pub fn from_reader(reader: &mut dyn Read) -> Result<Self> {
+        let mut state = hash_state();
+
+        let mut buf = [0u8; BUF_SIZE];
+        loop {
+            let bytes_read = reader.read(&mut buf)?;
+            if bytes_read == 0 {
+                break;
+            }
+            let chunk = &buf[0..bytes_read];
+            state.update(chunk);
+        }
+
+        Ok(state.finalize().into())
     }
 }
 
-impl TryFrom<Vec<u8>> for Blake2bHash {
-    type Error = Error;
-
-    // Create a Blake2bHash from a vector of bytes. Ensures the vector
-    // is the correct length for the hash.
-    fn try_from(value: Vec<u8>) -> Result<Self> {
-        let boxed: Box<[u8]> = value.into_boxed_slice();
-        let boxed: Box<[u8; HASH_DIGEST_SIZE]> =
-            boxed.try_into().map_err(|_| {
-                                 Error::CryptoError(format!("Did not get {} bytes for digest",
-                                                            HASH_DIGEST_SIZE))
-                             })?;
-        Ok(Blake2bHash::new(*boxed))
+// We *could* just wrap the `blake2b_simd::Hash` directly in our
+// `Blake2bHash` type, but then we wouldn't be able to parse a
+// `Blake2bHash` from a string, because `blake2b_simd::Hash does not
+// have that ability. So, we'll just grab the bytes and call it a
+// day.
+impl From<blake2b_simd::Hash> for Blake2bHash {
```

Comment:
You would still be able to implement `FromStr` for `Blake2bHash` even if it was wrapping `balke2b_simd::Hash` correct? Or am I misunderstanding this comment.

Reply:
That would turn into producing a `blake2b_simd::Hash` instance from the string, but I don't see a clear way to do that.